File: repos_estudiantes_TARQUI_tercera_entrega/SANCHEZ_SILVA_ALAN_PAULO/Generacidad/Ejercicio 3/ej3/src/Catalogo.java

import java.util.ArrayList;

public class Catalogo <T>{
    private ArrayList <T>  elementos;

    public Catalogo() {
        this.elementos = new ArrayList<>();
    }
    

    //a)    
    public void agregar(T elemento){
        this.elementos.add(elemento);
    }

    public boolean  buscar(T elemento){
        for (int i = 0; i < this.elementos.size() ; i++) {
            if(this.elementos.get(i).equals(elemento)){
                return true;
            }
        }
        return false;
    }
    
}

File: repos_estudiantes_TARQUI_tercera_entrega/RUBIN_DE_CELIS_VALDEZ_ANDRES/GENERECIDAD/EJERCICIO-3/catalogo.py
from typing import Generic, TypeVar, List, Callable

T = TypeVar('T')

class Catalogo(Generic[T]):
    def __init__(self):
        self.elementos: List[T] = []
    
    def agregar(self, elemento: T) -> None:
        self.elementos.append(elemento)
    
    def buscar(self, criterio: Callable[[T], bool]) -> List[T]:
        return [elem for elem in self.elementos if criterio(elem)]
    
    def __str__(self) -> str:
        return "\n".join(str(elem) for elem in self.elementos)
